### PR TITLE
added FailableCodable

### PR DESCRIPTION
**Pull request by oliverkrakora**

### CODE REVIEW

File: Sources/Toolbox/Utilities/FailableCodable.swift [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+/// The `websiteURL` is defined by a user so there is no guarantee that it is a valid `URL`.
+///
+///
+/// Without `FailableCodable` the decoding would fail for the entire `Person` and you would have to implement the decoding intializer yourself.
+///
+enum FailableCodable<T: Codable> {
+    /// Indicates that decoding of the underlying data was successfull
+    case success(T)
+    /// Indicates that the underlying data failed to decode
+    case error(Error)
```
Comment (swieland) [inline Sources/Toolbox/Utilities/FailableCodable.swift:28]: I would suggest to rename this case to `failure(Error)`

File: Sources/Toolbox/Utilities/FailableCodable.swift [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    
+    /// If the decoding was successfull a non nil value of type `T`
+    var value: T? {
+        switch self {
+        case .success(let value):
+            return value
+        case .error:
+            return nil
```
Comment (swieland) [inline Sources/Toolbox/Utilities/FailableCodable.swift:36]: switch can be replaces with:
```
guard case .error(let error) = self else { return nil }
```

File: Sources/Toolbox/Utilities/FailableCodable.swift [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            self = .error(error)
+        }
+    }
+    
+    func encode(to encoder: Encoder) throws {
+        var container = encoder.singleValueContainer()
+        if let value = value {
+            try container.encode(value)
+        } else {
+            try container.encodeNil()
```
Comment (swieland) [inline Sources/Toolbox/Utilities/FailableCodable.swift:53]: Decoding error case should throw an error. I think it's an implementation detail how to handle error case.

File: Sources/Toolbox/Utilities/FailableCodable.swift [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        do {
+            self = try .success(decoder.singleValueContainer().decode(T.self))
+        } catch {
+            self = .error(error)
+        }
+    }
+    
+    func encode(to encoder: Encoder) throws {
+        var container = encoder.singleValueContainer()
+        if let value = value {
```
Comment (swieland) [inline Sources/Toolbox/Utilities/FailableCodable.swift:50]: use switch case instead of if else

**Review by swieland — CHANGES_REQUESTED** [review]

File: Sources/Toolbox/Utilities/FailableCodable.swift (hunk 2 above)
Comment (mbuchetics) [reply Sources/Toolbox/Utilities/FailableCodable.swift:36]: finde den switch hier aber besser (lesbarer)

File: Sources/Toolbox/Utilities/FailableCodable.swift (hunk 3 above)
Comment (mbuchetics) [reply Sources/Toolbox/Utilities/FailableCodable.swift:53]: hm, I wouldn't throw an error. Imagine you fail to decode some property, resulting into an error case (that's what the whole FailableCodable is for) and then you want to store it, send back to the server, etc. This shouldn't throw an error. So I think the current behaviour is correct.

**Review by mbuchetics — APPROVED** [review]

File: Sources/Toolbox/Utilities/FailableCodable.swift (hunk 3 above)
Comment (wieweb) [reply Sources/Toolbox/Utilities/FailableCodable.swift:53]: You are right, this is the purpose of FailableCodable ;-)

**Pull request merged**